Add database

PaperAlex opened the pull request.

PaperAlex committed:
*Техническое задание проекта №10 1.0
1. Во процессе выполнения добавил доп задание jakarta.validation

PaperAlex committed:
*Техническое задание проекта №10 1.0
1. Во процессе выполнения добавил доп задание jakarta.validation

PaperAlex committed:
*Техническое задание проекта №10 1.2
1. Обновить логирование

PaperAlex committed:
*Техническое задание проекта №10 1.3
1. Обновить логирование
2. Добавить apache StringUtils
3. Исправить исключения extends Exception
4. Создать метод для тестов

PaperAlex committed:
*Техническое задание проекта №10 1.3
1. Обновить логирование
2. Добавить apache StringUtils
3. Исправить исключения extends Exception
4. Создать метод для тестов

PaperAlex committed:
*Техническое задание проекта №11 1.0

PaperAlex committed:
Merge branch 'main' into add-friends-likes

PaperAlex committed:
*Техническое задание проекта №11 1.0

PaperAlex committed:
Merge remote-tracking branch 'origin/add-friends-likes' into add-friends-likes

PaperAlex committed:
*Техническое задание проекта №11 1.0

PaperAlex committed:
*Техническое задание проекта №11 1.0

PaperAlex committed:
*Техническое задание проекта №11 1.0

PaperAlex committed:
*Техническое задание проекта №11 1.1
Исправить тесты

PaperAlex committed:
*Техническое задание проекта №11 1.2
Исправить Service и Storage

PaperAlex committed:
*Техническое задание проекта №11 1.2
Исправить Service и Storage

PaperAlex committed:
*Техническое задание проекта №11 1.2
Исправить Service и Storage

PaperAlex committed:
*Техническое задание проекта №11 1.2
Исправить Service и Storage

PaperAlex committed:
*Промежуточное задание проекта №12

PaperAlex committed:
*Промежуточное задание проекта №12

PaperAlex committed:
Merge branch 'main' into add-friends-likes

PaperAlex committed:
*Промежуточное задание проекта №12

PaperAlex committed:
Merge remote-tracking branch 'origin/add-friends-likes' into add-friends-likes

PaperAlex committed:
*Промежуточное задание проекта №12

PaperAlex committed:
JDBC установить драйвер и создать таблицу через java

PaperAlex committed:
JDBC установить драйвер pom

PaperAlex committed:
*Техническое задание проекта №12 1.0

PaperAlex committed:
Merge branch 'main' into add-database

PaperAlex committed:
*Техническое задание проекта №12 1.0

PaperAlex committed:
Merge remote-tracking branch 'origin/add-database' into add-database

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,32 +1,43 @@
     @PostMapping
-    public Film create(@Valid @RequestBody Film film) throws ValidationException {
-       return filmService.create(film);
+    public Film create(@Valid @RequestBody Film film) throws NotFoundException, ValidationException, SQLException {
```
SQLException

Вот не уверен что стоит так высоко прямо SQLException прокидывать, по сути дела мы в контракте вэб-сервиса декларируем пользователям (например, другим сервисам) нашего сервиса, что что надо обрабатывать именно такую ошибку. По хорошему эту бы ошибку так высоко не кидать, перехватывать (обрабатывать или заменять другой, уже бизнесовым исключением) на уровне сервиса или даже хранилища.

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,53 +1,69 @@
-        Optional<User> user = userStorage.findById(userId);
+        Optional<Film> film = filmStorage.findFilmById(filmId);
+        Optional<User> user = userStorage.findUserById(userId);
         if (film.isEmpty()) {
```
film.isEmpty()

С Optional лучше вот так сделать, в более функциональном стиле

Film film = storage.findUser(....).orElseThrow(new NotFoundException(...))

https://struchkov.dev/blog/ru/optional-in-java/

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,53 +1,69 @@
     }
 
-    public Collection<Film> popularFilms(Integer count) throws ValidationException {
+    public Collection<Film> findPopularFilms(Integer count) throws ValidationException {
```
А через аннотации не получается сделать проверку на null, что то типа @NotNull и пр :)
На отрицательнео число кстати будет такая же ошибка, про null.

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -59,26 +75,24 @@ public Collection<Film> findAll() {
 
     public Film update(Film newFilm) throws ValidationException, NotFoundException {
-        if (newFilm.getId() == null) {
+        if (filmStorage.findFilmById(newFilm.getId()).isEmpty()) {
```
Можно кстати вот такой метод сделать

filmStorage.existsFilmById(newFilm.getId()): boolean

Можно попробовать вот такой запрос прикрутить, чутка эффективней будет и сам код наглядней 

https://stackoverflow.com/questions/7471625/fastest-check-if-row-exists-in-postgresql

EugeneLenkevich replied:
Чет слишком сложный запрос, предлагал только один метод заменить - filmStorage.existsFilmById :)

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -94,4 +108,11 @@ private void releaseDateValidation(LocalDate releaseDate) throws ValidationExcep
     }
+
+    private void mpaValidation(Integer mpaId) throws ValidationException {
+        Integer mpaCount = mpaStorage.findAllMpa().size();
```
mpaStorage.findAllMpa()

Мы тут все записи из таблицы вытаскиваем, это может быть опасно, если записей много, ну и не особо эффестивно. Однако, мы же можем запросом все это посчитать 

select count(id) from mpa

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/GenreService.java:
```diff
@@ -0,0 +1,38 @@
+
+    public Optional<Genre> findGenreById(int id) throws NotFoundException {
+        Optional<Genre> genre = genreStorage.findGenreById(id);
+        if (genre.isPresent()) {
```
Опять таки, это, с Optional, можно записать покороче. И посмотри, пожалуйста, аналогичные места :)

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,47 +1,54 @@
     public void addFriend(Long userId, Long friendId) throws NotFoundException {
-        findUserById(userId).getFriends().add(friendId);
-        findUserById(friendId).getFriends().add(userId);
+        if (userStorage.findUserById(userId).isEmpty() || userStorage.findUserById(friendId).isEmpty()) {
```
Вот с таким бы методом все бы стало покороче :)
userStorage.existsById(userId)

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -56,20 +63,31 @@ public Collection<User> findAll() {
     }
 
+    private void userNameValidation(User user) {
+        if (user.getName() == null || user.getName().isBlank()) {
```
Для проверки строк на пустоту можно использовать 
StringUtils.hasText
или же аналогичную функцию из apache commons 
StringUtils.isNotBlank
=)

EugeneLenkevich reviewed (changes requested):
Привет, Алексей! Спасибо за PR.

Отличная работа! :)
Есть совсем небольшие пометки, посмотри комментарии ниже, пожалуйста

Основные моменты 

- попробуй упросить код с Optional, синтаксический сахар, его цель сделать код проще для чтения, посмотрим на сколько он короче станет
- если получится то exist-запросы сделать бы, ну вообще, чем мы больше обработки ( имею в виду выборку, фильтрацию и пр данных) перекладываем на СУБД тем лучше, так как СУБД это делает более эффективно, но без фанатизма, главное постараться не тащить лишние данные :)

PaperAlex committed:
*Техническое задание проекта №12 1.2
1. Исправлены исключение и Optional
2. Запрос count mpa
3. Добавлены exists методы

PaperAlex committed:
*Техническое задание проекта №12 1.2
1. Исправлены исключение и Optional
2. Запрос count mpa
3. Добавлены exists методы

PaperAlex committed:
*Техническое задание проекта №12 1.2
1. Исправлены исключение и Optional
2. Запрос count mpa
3. Добавлены exists методы

PaperAlex committed:
*Техническое задание проекта №12 1.2
1. Исправлены исключение и Optional
2. Запрос count mpa
3. Добавлены exists методы

PaperAlex committed:
*Техническое задание проекта №12 1.2
1. Исправлены исключение и Optional
2. Запрос count mpa
3. Добавлены exists методы

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/UserDbStorage.java:
```diff
@@ -126,4 +126,16 @@ public List<User> findMutualFriends(Long userId, Long friendId) {
+
+    @Override
+    public boolean existById(Long userId, Long friendId) throws NotFoundException {
+        if (findUserById(userId).isEmpty() || findUserById(friendId).isEmpty()) {
```
findUserById(userId)
findUserById(friendId)

Вот к этим двум методам не получилось сделать аналогичную пару exists-методов с такими примерно запросами, тут не надо дополнительных сложностей, именно такие простые запросы ? 

`SELECT  EXISTS(SELECT 1 FROM user WHERE id = :userId);`

PaperAlex replied:
Кажется разобрался)
`public boolean existFilmById(Long filmId) throws NotFoundException {
        String sql = "SELECT EXISTS(SELECT 1 FROM films WHERE film_id = ?);";
        if(Boolean.FALSE.equals(jdbcTemplate.queryForObject(sql, Boolean.class, filmId))) {
            throw new NotFoundException("Должен быть указан существующий id");
        }
        return true;
    }`

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/MpaService.java:
```diff
@@ -27,13 +27,13 @@ public List<Mpa> findAllMpa() {
-        } else {
-            throw new NotFoundException("Rating does not exist");
-        }
+        Mpa mpa = mpaStorage.findMpaById(id).orElseThrow(() -> new NotFoundException("Rating does not exist"));
```
Часто подобное, как и другие стримы, разбивают на несколько строк, для удобства чтения, а то слишком много всего в строчке получается
```
Mpa mpa = mpaStorage.findMpaById(id)
    .orElseThrow(() -> new NotFoundException("Rating does not exist"));
```

EugeneLenkevich reviewed (approved):
Ага, ок, пусть будет так, inmemory-реализацию можно было просто удалить чтобы не мешалась, если что в гите она есть 
Потом у нас будут спринговые репозитории, с ними запросы писать гораздо удобней, но общие принципы те же самые :)

PaperAlex committed:
*Техническое задание проекта №12 1.2
1. Исправлены исключение и Optional
2. Запрос count mpa
3. Добавлены exists с запросами sql

PaperAlex committed:
*Техническое задание проекта №12 1.2
1. Исправлены исключение и Optional
2. Запрос count mpa
3. Добавлены exists с запросами sql
